Java/jni/native_glue.cc

```diff
@@ -84,6 +92,74 @@ namespace litecore {
                 free((void *) _slice.buf);        // detached
         }
 
+        void jstringSlice::UTF8ToModifiedUTF8(const char *input, char *output) {
+            int unicodePoint = (input[0]-240<<18) +
+                               (input[1]-128<<12) +
+                               (input[2]-128<<6) +
+                               (input[3]-128) - 0x10000;
+
+            uint16_t surrogates[2] { (uint16_t)0xD800 | (uint16_t)(unicodePoint >> 10),
+                                     (uint16_t)0xDC00 | (uint16_t)(unicodePoint & 0x3FF) };
+
+            output[0] = surrogates[0]>>12 & 0x0F | 0xE0;
+            output[1] = surrogates[0]>>6 & 0x3F | 0x80;
+            output[2] = surrogates[0] & 0x3F | 0x80;
+            output[3] = surrogates[1]>>12 & 0x0F | 0xE0;
+            output[4] = surrogates[1]>>6 & 0x3F | 0x80;
+            output[5] = surrogates[1] & 0x3F | 0x80
+        }
+
+        void jstringSlice::fixBrokenModifiedUTF8(const char *input){
+            // https://github.com/android-ndk/ndk/issues/283
+            size_t newStrLen = 0;
+            size_t inputByteCount = 0;
+            const auto unsignedInput = (const uint8_t *)input;
+
+            // Need to figure out the actual length since each
+            // 4-bytes of real UTF-8 is 6 bytes of modified UTF-8
+            // but convert the bytes while we are at it
+            for(size_t i = 0; unsignedInput[i] != 0; i++) {
+                if(unsignedInput[i] >= 0xF0) {
+                    // 0xF0 and above marks 4-byte sequences
+                    newStrLen += 2;
+
+                    // 3 + 1 from next loop iteration = 4 bytes advance
+                    i += 3;
+                }
+
+                newStrLen++;
+                inputByteCount++;
```

Comment:
Isn't inputByteCount the same as i?

Reply:
Oh wait, it's not, because inputByteCount doesn't include 3 of the bytes from the 4-byte sequences.
I think that means the second `for` loop will go wrong...